lib/saml/auth_fail_handler.rb

```diff
@@ -17,15 +19,17 @@ def initialize(saml_response, user, session)
       @context = nil
     end
 
-    def known_error?
-      known_errors = [
-        :clicked_deny?,
-        :auth_too_late?,
-        :auth_too_early?
-      ]
+    def error
+      KNOWN_ERRORS.each do |known_error|
+        return known_error if send("#{known_error}?")
+      end
```

Comment:
so if you have the hash of the error types:

{ :error_type => 'error message', :error_type2 => 'error_message' }

You can transpose this hash so that it is: 'error message' => :error_type

using Hash#transpose

Then you just compare the JSON response message with the hash and you know exactly what the method is, and it is O(1) time vs O(n) time (looping through known_errors. And the logic around whether or not it is a single error or not could be handled therein as well.